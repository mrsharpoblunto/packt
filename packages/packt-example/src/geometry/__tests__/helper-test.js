import glm from 'gl-matrix';
import {
  closestPointLineSegments,
  closestPointOnTriangle
} from 'geometry/helpers';

describe('closest point on triangle', () => {
  it('finds the closest point to a triangle', () => {
    const out = glm.vec3.create();
    const p = glm.vec3.fromValues(0, 1, 0);
    const a = glm.vec3.fromValues(1, 0, 1);
    const b = glm.vec3.fromValues(-1, 0, 1);
    const c = glm.vec3.fromValues(0, 0, -1);
    expect([closestPointOnTriangle(out, p, a, b, c), out]).toMatchSnapshot();
  });

  it('finds the closest point to a triangle edge', () => {
    const out = glm.vec3.create();
    const p = glm.vec3.fromValues(0, 0, 2);
    const a = glm.vec3.fromValues(1, 0, 1);
    const b = glm.vec3.fromValues(-1, 0, 1);
    const c = glm.vec3.fromValues(0, 0, -1);
    expect([closestPointOnTriangle(out, p, a, b, c), out]).toMatchSnapshot();
  });

  it('finds the closest point touching a triangle', () => {
    const out = glm.vec3.create();
    const p = glm.vec3.fromValues(0, 0, -0.5);
    const a = glm.vec3.fromValues(1, 0, 1);
    const b = glm.vec3.fromValues(-1, 0, 1);
    const c = glm.vec3.fromValues(0, 0, -1);
    expect([closestPointOnTriangle(out, p, a, b, c), out]).toMatchSnapshot();
  });
});

describe('closest point between line segments', () => {
  it('finds the closest point between two points', () => {
    const outA = glm.vec3.create();
    const outB = glm.vec3.create();

    const a1 = glm.vec3.fromValues(0.5, 0, 0);
    const a2 = glm.vec3.fromValues(0.5, 0, 0);
    const b1 = glm.vec3.fromValues(0, 0.5, 0);
    const b2 = glm.vec3.fromValues(0, 0.5, 0);

    expect([
      // the value returned is the distance squared
      closestPointLineSegments(outA, outB, a1, a2, b1, b2),
      outA,
      outB
    ]).toMatchSnapshot();
  });

  it('finds the closest point between two touching segments', () => {
    const outA = glm.vec3.create();
    const outB = glm.vec3.create();

    const a1 = glm.vec3.fromValues(0.5, 0, 0);
    const a2 = glm.vec3.fromValues(2, 1, 0);
    const b1 = glm.vec3.fromValues(0, 0.5, 0);
    const b2 = glm.vec3.fromValues(2, 1, 0);

    expect([
      closestPointLineSegments(outA, outB, a1, a2, b1, b2),
      outA,
      outB
    ]).toMatchSnapshot();
  });

  it('finds the closest point between two intersecting segments', () => {
    const outA = glm.vec3.create();
    const outB = glm.vec3.create();

    const a1 = glm.vec3.fromValues(0.5, 0, 0);
    const a2 = glm.vec3.fromValues(2, 1, 0);
    const b1 = glm.vec3.fromValues(0, 0.5, 0);
    const b2 = glm.vec3.fromValues(2, 0.5, 0);

    expect([
      closestPointLineSegments(outA, outB, a1, a2, b1, b2),
      outA,
      outB
    ]).toMatchSnapshot();
  });

  it('finds the closest point between non-intersecting segments', () => {
    const outA = glm.vec3.create();
    const outB = glm.vec3.create();

    const a1 = glm.vec3.fromValues(0, 0, 0);
    const a2 = glm.vec3.fromValues(2, 1, 0);
    const b1 = glm.vec3.fromValues(1, 0, 0);
    const b2 = glm.vec3.fromValues(1.5, 0.5, 0);

    expect([
      closestPointLineSegments(outA, outB, a1, a2, b1, b2),
      outA,
      outB
    ]).toMatchSnapshot();
  });
});
