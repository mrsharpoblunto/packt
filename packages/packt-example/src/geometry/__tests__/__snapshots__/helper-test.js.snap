exports[`closest point between line segments finds the closest point between non-intersecting segments 1`] = `
Array [
  0.05000000000000071,
  Float32Array [
    1.399999976158142,
    0.699999988079071,
    0
  ],
  Float32Array [
    1.5,
    0.5,
    0
  ]
]
`;

exports[`closest point between line segments finds the closest point between two intersecting segments 1`] = `
Array [
  0,
  Float32Array [
    1.25,
    0.5,
    0
  ],
  Float32Array [
    1.25,
    0.5,
    0
  ]
]
`;

exports[`closest point between line segments finds the closest point between two points 1`] = `
Array [
  0.5,
  Float32Array [
    0.5,
    0,
    0
  ],
  Float32Array [
    0,
    0.5,
    0
  ]
]
`;

exports[`closest point between line segments finds the closest point between two touching segments 1`] = `
Array [
  0,
  Float32Array [
    2,
    1,
    0
  ],
  Float32Array [
    2,
    1,
    0
  ]
]
`;

exports[`closest point on triangle finds the closest point to a triangle 1`] = `
Array [
  1,
  Float32Array [
    0,
    0,
    0
  ]
]
`;

exports[`closest point on triangle finds the closest point to a triangle edge 1`] = `
Array [
  1,
  Float32Array [
    0,
    0,
    1
  ]
]
`;

exports[`closest point on triangle finds the closest point touching a triangle 1`] = `
Array [
  0,
  Float32Array [
    0,
    0,
    -0.5
  ]
]
`;
